File: questions/1717. Maximum Score From Removing Substrings /maximum.java
/*
You are given a string s and two integers x and y. You can perform two types of operations any number of times.

Remove substring "ab" and gain x points.
For example, when removing "ab" from "cabxbae" it becomes "cxbae".
Remove substring "ba" and gain y points.
For example, when removing "ba" from "cabxbae" it becomes "cabxe".
Return the maximum points you can gain after applying the above operations on s.
*/
class Solution {
    public int maximumGain(String s, int x, int y) {
        char first = (x > y)?'a':'b';
        char second = (x > y)?'b':'a';
        int max = Math.max(x, y);
        int min = Math.min(x, y);
        int res = 0;
        Stack<Character> stack1 = new Stack();
        Stack<Character> stack2 = new Stack();
        for(char c: s.toCharArray()){
            if(!stack1.isEmpty() && stack1.peek() == first && c == second){
                stack1.pop();
                res += max;
            }else{
                stack1.push(c);
            }
        }
        while(!stack1.isEmpty()){
            char c = stack1.pop();
            if(!stack2.isEmpty() && stack2.peek() == first && c == second){
                stack2.pop();
                res += min;
            }else{
                stack2.push(c);
            }
        }
        return res;
    }
}

File: questions/1793. Maximum Score of a Good Subarray/maximum.py
'''
You are given an array of integers nums (0-indexed) and an integer k.

The score of a subarray (i, j) is defined as min(nums[i], nums[i+1], ..., nums[j]) * (j - i + 1). A good subarray is a subarray where i <= k <= j.

Return the maximum possible score of a good subarray.
'''
class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        res, min_num = nums[k], nums[k]
        start, end, n = k, k, len(nums)
        while start > 0 or end < n - 1:
            if (nums[start - 1] if start else 0) >= (nums[end + 1] if end < n - 1 else 0):
                start = start - 1
            else:
                end = end + 1
            min_num = min(min_num, nums[start], nums[end])
            res = max(res, min_num * (end - start + 1))
        return res